# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB

4
5
9

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
4

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

4
Traceback (most recent call last):
  File "sum.py", line 7, in <module>
    a, b = map(int, input().split())
ValueError: not enough values to unpack (expected 2, got 1)


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

 5 10 100 1000 10000 100000
1000000000


### Test 1: Small numbers

In [2]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [3]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Writing input2.txt


In [ ]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [ ]:
%%writefile generador.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>

using namespace std;

int main() {
  int n;
  int aleatorio;
  srand((unsigned) time(0));
  ofstream MyFile("stress.txt");
  n = 2 + ( rand() % ( 20 - 2 + 1 ) );
  MyFile << n << "\n";
  for(int i = 0; i < n; i++){
    aleatorio = rand() % 200000;
    MyFile << aleatorio << "  ";
  }
  cout << "\n";
  // Close the file
  MyFile.close();
  return 0;
}

Overwriting generador.cpp


In [ ]:
!g++ generador.cpp -o generador
import time
for i in range (20):
  !./generador
  time.sleep(1)
  !./max_pairwise_product_long < stress.txt


19457416623

2903101131

10253512395

31378506840

12843876120

30826978298

9980288835

31754638200

30243672344

22036579370

13850137920

33617692508

25327339452

9482085045

32673772012

16714298700

22637633551

8461472162

9002777668

32685906045


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [8]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Writing Test2Write.cpp


In [9]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

39999800000

real	2m7.759s
user	2m7.376s
sys	0m0.020s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182

2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

---

---


### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile efficient_maxpairwise.cpp
#include <iostream>
#include <vector>
using namespace std;
long long MaxPairwiseProductFast(const vector<int>& numbers){
    int n = numbers.size();

    int max_index1 = -1;
    for(int i = 0; i<n ; i++){      
        if((max_index1 == -1 )|| (numbers[i]> numbers[max_index1])){
            max_index1 = i;
        }
    }
    int max_index2 = -1;
    for(int j = 0; j<n ; j++){
        if((j != max_index1) && ((max_index2 == -1) || (numbers[j] > numbers[max_index2]))){
            max_index2 = j;
        }
    }
    return ((long long)(numbers[max_index1])) * numbers[max_index2];
}
int main(){ 
    int n; 
    cin >> n;
    vector<int> arr(n);
    for(int i = 0; i < n; i++ ){
        cin >> arr[i];
    }
    cout << MaxPairwiseProductFast(arr)<<"\n";
    
    return 0;
}

Writing efficient_maxpairwise.cpp


Test with inputs 1 and 2



In [ ]:
!g++ efficient_maxpairwise.cpp -o efficient_maxpairwise
!./efficient_maxpairwise < input1.txt
!./efficient_maxpairwise < input2.txt

90
40000000000


In [ ]:
!time ./efficient_maxpairwise < input1.txt
!time ./efficient_maxpairwise < input2.txt

90

real	0m0.003s
user	0m0.001s
sys	0m0.002s
40000000000

real	0m0.003s
user	0m0.002s
sys	0m0.001s


Test and time it with input 3

In [ ]:
!time ./efficient_maxpairwise < input3.txt

39999800000

real	0m0.060s
user	0m0.059s
sys	0m0.001s


20 stress tests

In [ ]:
import time
for i in range (20):
  !./generador
  time.sleep(1)
  !./efficient_maxpairwise < stress.txt


23691309

35959586773

27940903092

134695524

25925947213

31724618010

34917636789

37008150244

10011462960

18925369407

36551147709

15614784215

34035799486

33100971898

31312002990

28016790908

30920586717

34282549749

9223619712

34787890704


Count executed instructions in code and verify with inputs 1,2 and 3


Overwriting efficient_maxpairwiseCount.cpp


In [ ]:
!g++ MaxPairwiseProductFastCount.cpp -o MaxPairwiseProductFastCount
!./MaxPairwiseProductFastCount < input1.txt
!./MaxPairwiseProductFastCount < input2.txt
!./MaxPairwiseProductFastCount < input3.txt

g++: error: MaxPairwiseProductFastCount.cpp: No such file or directory
g++: fatal error: no input files
compilation terminated.
/bin/bash: ./MaxPairwiseProductFastCount: No such file or directory
/bin/bash: ./MaxPairwiseProductFastCount: No such file or directory
/bin/bash: ./MaxPairwiseProductFastCount: No such file or directory


Calculate estimated time by instruction and number of instructions by second

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [ ]:
%%writefile superfast.cpp
#include <iostream>

using namespace std;

int main() {
    long long n, a = 0, b = 0, num = 1;

    cin >> n;
    
    for (int i = 0; i < n; i++){
        cin >> num;
        if(num >= a){
            b = a;
            a = num;
        }
        else if(num > b)
            b = num;
    }

    cout << a*b<<"\n";

return 0;
}

Writing superfast.cpp


Test with inputs 1 and 2


 

In [ ]:
!g++ superfast.cpp -o superfast
!./superfast < input1.txt
!./superfast < input2.txt

90
40000000000


In [ ]:
!time ./superfast < input1.txt
!time ./superfast < input2.txt

90

real	0m0.003s
user	0m0.001s
sys	0m0.002s
40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


Test and time it with input 3

In [ ]:
!time ./superfast < input3.txt

39999800000

real	0m0.062s
user	0m0.055s
sys	0m0.004s


20 stress tests 

In [ ]:
import time
for i in range (20):
  !./generador
  time.sleep(1)
  !./superfast < stress.txt


37209087902

33043416420

33037656384

34822979944

27654943668

23638170010

32483934240

37423405260

16733683794

20548496632

25579709120

33279489920

22889424470

33252397290

36595135694

21769596491

22508686086

15910471676

19086780543

36135757770


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile superfastCount.cpp
#include <iostream>

using namespace std;

int main() {
    
    long long execints =0LL;
    
    long long n, a = 0, b = 0, num = 1;  //1 + 1 + 1 + 1 
    execints += 4;
    
    cin >> n;                            //1 
    execints++;
    
    for (int i = 0; i < n; i++){         //1+2n 
        execints += 2;
        cin >> num;                      // 1
        execints++;                     
        if(num >= a){                    //(n-1)n/2                 
            b = a;
            execints++;
            a = num;  
            execints++;                //[0, (n-1)n/2] 
        }
        else if(num > b){               //(n-1)n/2
            b = num;                    
            execints++;                 //[0, (n-1)n/2]
        }
    }                                   //n  
    cout << a*b<<"\n";                  //1
    execints++;
    cout <<"lineas ejecutadas: "<< execints<<"\n";
return 0;
}

Writing superfastCount.cpp


In [ ]:
!g++ superfastCount.cpp -o superfastCount
!./superfastCount < input1.txt
!./superfastCount < input2.txt
!./superfastCount < input3.txt

90
lineas ejecutadas: 56
40000000000
lineas ejecutadas: 33
39999800000
lineas ejecutadas: 1000006


$$T(n)=6+5*(n)=6+5*(10)=56$$

$$T(n)=6+5(n-1)+2=33$$

Calculate estimated time by instruction and number of instructions by second

1-$$\text{Tiempo por instruccion: }\frac{0.004}{56}=7.14\times 10^{-5}s$$
  $$\text{Instrucciones por segundo: }\frac{1}{\frac{0.004}{56}}=14000$$
2-$$\text{Tiempo por instruccion: }\frac{0.003}{33}=9.09\times10^{-5}s$$
  $$\text{Instrucciones por segundo: }\frac{1}{\frac{0.003}{33}}=11000$$
3-$$\text{Tiempo por instruccion: }\frac{0.063}{1000006}=6.3\times10^{-8}s$$
$$\text{Instrucciones por segundo: }\frac{1}{\frac{0.063}{1000006}}=15873111$$

## Python

Efficient cython file reading, string parsing, and array building

https://stackoverflow.com/questions/37219875/efficient-cython-file-reading-string-parsing-and-array-building



File PreProcessing With C and Cython

https://github.com/EveryTimeIWill18/Cython_Repo/blob/master/FastFileProcessingWithCython.ipynb



#Efficient solution python

In [5]:
%%writefile pythoneff.py
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))

Overwriting pythoneff.py


Test 1 and Test 2

In [7]:
!python pythoneff.py < input1.txt
!python pythoneff.py < input2.txt

90
40000000000


Measuring time with test 3

In [11]:
!time python pythoneff.py < input3.txt

Traceback (most recent call last):
  File "pythoneff.py", line 15, in <module>
    print(max_pairwise_product(input_numbers))
  File "pythoneff.py", line 7, in max_pairwise_product
    numbers[first] * numbers[second])
KeyboardInterrupt

real	6m22.671s
user	6m21.055s
sys	0m0.107s


# Screenhsoots of the Coursera Submissions in C++ and Python

![](https://drive.google.com/uc?id=1jD8ZedSpbO_ar5x-GnM_fJReeUvfvrg3)

![](https://drive.google.com/uc?id=1LSIlLyYbhdfQvVBDsW_wsXdlS-N59MgW)